README.md

```diff
@@ -1,44 +1,284 @@
+    // filter_action
+    SeccompAction::Allow,
+    // target architecture of filter
+    TargetArch::x86_64,
+    // rule set
+    vec![
+        (
+            libc::SYS_accept4,
+            vec![]
+        ),
+        (
+            libc::SYS_fcntl,
+            vec![
+                SeccompRule::new(vec![
+                    Cond::new(1,
+                        SeccompCmpArgLen::DWORD,
+                        SeccompCmpOp::Eq,
+                        libc::F_SETFD
+                    )?,
+                    Cond::new(
+                        2,
+                        SeccompCmpArgLen::DWORD,
+                        SeccompCmpOp::Eq,
+                        libc::FD_CLOEXEC,
+                    )?,
+                ]),
+                SeccompRule::new(vec![
+                    Cond::new(
+                        1,
+                        SeccompCmpArgLen::DWORD,
+                        SeccompCmpOp::Eq,
+                        libc::F_GETFD,
+                    )?
+                ])
+            ]
+        )
+    ]
+)?
+```
+
+## Example usage
+
+Using seccompiler in an application is a two-step process:
+
+1. Compiling filters (into BPF)
+2. Installing filters
+
+### Compiling filters
+
+A user application can compile the seccomp filters into loadable BPF either at
+**runtime** or at **build time**.
+
+At **runtime**, the process is straightforward, leveraging the seccompiler
+library functions on hardcoded/file-based filters.
+
+At **build-time,** an application can use a
+[cargo build script](https://doc.rust-lang.org/cargo/reference/build-scripts.html)
+that adds seccompiler as a build-dependency and outputs at a predefined
+location (e.g. using `env::var("OUT_DIR")`) the compiled filters, that have
+been serialized to a binary format
+(e.g. [bincode](https://github.com/bincode-org/bincode)).
+They can then be ingested by the application using `include_bytes!` and
+deserialized before getting installed.
+This build-time option can be used to shave off the filter compilation time
+from the app startup time, if using a low-overhead binary format.
+
+Regardless of the compilation moment, the process is the same:
+
+For **JSON filters**, the compilation to loadable BPF is performed using the
+`compile_from_json()` function:
 
 ```rust
-use my_crate;
+let filters: BpfMap = seccompiler::compile_from_json(
+    File::open("/path/to/json")?, // Accepts generic Read objects.
+    seccompiler::TargetArch::x86_64,
+)?;
+```
 
-...
+`BpfMap` is another type exposed by the library, which maps thread
+categories to BPF programs.
+
+```rust
+pub type BpfMap = HashMap<String, Arc<BpfProgram>>;
 ```
 
-## License
+For **Rust filters**, it’s enough to perform a `try_into()` cast, from a
+`SeccompFilter`, to a `BpfProgram`, like so:
+
+```rust
+let seccomp_filter = SeccompFilter::new(
+    rule_map,
+    SeccompAction::Trap,
+    seccompiler::TargetArch::x86_64
+)?;
+
+let bpf_prog: BpfProgram = seccomp_filter.try_into()?;
```

Comment:
Is this the format expected by the kernel?

Reply:
yes, it's a Vec<sock_filter> and the kernel expects a `sock_filter` pointer